testdata/generate.py

```diff
@@ -0,0 +1,526 @@
+            )
+
+        self.writer.writerow(attrs)
+        return attrs
+
+
+class RecurringDonationGenerator(Generator):
+    table = "recurring_donations"
+    columns = ("id", "contact_id", "name", "amount", "start_date", "open_ended")
+
+    def __init__(self, conn, logger, distribution, opportunity_generator):
+        super(RecurringDonationGenerator, self).__init__(conn, logger, distribution)
+        self.generators["opportunities"] = opportunity_generator
+
+    def make_one(self, account, contacts, **kw):
+        attrs = {}
+        attrs["id"] = id = self.make_id()
+        contact = contacts[0]
+        attrs["contact_id"] = contact["id"]
+        attrs["name"] = "{} {} Recurring Donation".format(
+            contact["first_name"], contact["last_name"]
+        )
+        attrs["amount"] = amount = random.randrange(25, 225, 25)
+        year = random.randrange(this_year - 10, this_year + 1)
+        attrs["start_date"] = "{}-01-01".format(year)
+        attrs["open_ended"] = year == this_year
+        attrs.update(kw)
+
+        opp_gen = self.generators["opportunities"]
+        for month in MONTHS:
+            is_past = year < this_year or month < this_month
+            opp_gen.make_one(
+                contacts,
+                account_id=account["id"],
+                contact_id=contact["id"],
+                recurring_donation_id=id,
+                amount=amount,
+                stage_name="Closed Won" if is_past else "Pledged",
+                close_date="{}-{:02d}-01".format(year, month),
+                dont_create_payments=False,
+            )
+
+        self.writer.writerow(attrs)
+        return attrs
+
+
+class HouseholdGenerator(Generator):
+    table = "accounts"
+    columns = ("id", "name", "record_type")
+
+    def __init__(self, conn, logger, include_skew=False):
+        super(HouseholdGenerator, self).__init__(conn, logger)
+        self.include_skew = include_skew
+
+    def init_subgenerators(self):
+        opportunity_generator = OpportunityGenerator(
+            self.conn,
+            self.logger,
+            {0: 2, 1: 3, 2: 2, 3: 2, 4: 1, 5: 1, 6: 1, 7: 1, 8: 1, 9: 1, 10: 1},
+        )
+        self.generators = {
+            "contacts": ContactGenerator(
+                self.conn, self.logger, {1: 10, 2: 10, 3: 1, 4: 1, 5: 1}
+            ),
+            "addresses": AddressGenerator(self.conn, self.logger, {1: 50, 2: 1}),
+            "recurring_donations": RecurringDonationGenerator(
+                self.conn, self.logger, {1: 1, 0: 4}, opportunity_generator
+            ),
+            "opportunities": opportunity_generator,
+        }
+
+    def make_one(self, n):
+        id = self.make_id()
+        attrs = {"id": id, "record_type": "HH_Account"}
+
+        # contacts
+        contact_gen = self.generators["contacts"]
+        num_contacts = contact_gen.choose_count()
+        contacts = []
+        for x in range(num_contacts):
```

Comment:
Why not use a list comprehension here? I thought they were faster than appending?

Reply:
I tend to avoid list comprehensions for readability reasons, but I think it's a good call here.